[view-transitions] Implement matching for view transition classes

nt1m opened the pull request:
#### 28a8cb798f0cf5d11677902f6f8f87820f6bb286
<pre>
[view-transitions] Implement matching for view transition classes
<a href="https://bugs.webkit.org/show_bug.cgi?id=278223">https://bugs.webkit.org/show_bug.cgi?id=278223</a>
<a href="https://rdar.apple.com/134020027">rdar://134020027</a>

Reviewed by Darin Adler.

Implements matching as described in:
<a href="https://drafts.csswg.org/css-view-transitions-2/#shared-style-with-vt-classes">https://drafts.csswg.org/css-view-transitions-2/#shared-style-with-vt-classes</a>

* LayoutTests/TestExpectations:

Unskip all tests.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:

Enable by default, since all WPT pass.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/css/CSSSelector.h:

Add PossiblyQuotedIdentifier TextStream for easy debugging.

* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorChecker.h:

All classes in need to match in order for the styles to apply.

* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::effectiveViewTransitionClassList):
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::captureNewState):

Captures the classList as described in <a href="https://drafts.csswg.org/css-view-transitions-2/#vt-class-algorithms">https://drafts.csswg.org/css-view-transitions-2/#vt-class-algorithms</a>

* Source/WebCore/dom/ViewTransition.h: Add classList to CapturedElement.

* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::pseudoStyleCacheIsInvalid):

Fix a bug in the cache where were passing a null name argument string.

* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::classListForNamedViewTransitionPseudoElement):
(WebCore::Style::ElementRuleCollector::ruleMatches):

Pass the classList to the SelectorChecker when checking the selector.

Canonical link: <a href="https://commits.webkit.org/282383@main">https://commits.webkit.org/282383@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3f4800cada96ae64f5e6591773e11fd9b7b9b848

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63018 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42374 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15614 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/67039 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/13622 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65138 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50061 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13906 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/67039 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/13622 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66087 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/39365 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/67039 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/36050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/11916 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/12498 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/56128 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/57593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/12245 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/68734 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/62261 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/6964 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Debug-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/68734 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/6996 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/54641 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/68734 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/84024 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/38194 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [⏳ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Tests-EWS "Waiting to run tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/39274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/40385 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/39016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSSelector.h:
```diff
@@ -26,6 +26,7 @@
 #include "RenderStyleConstants.h"
 #include <wtf/EnumTraits.h>
 #include <wtf/FixedVector.h>
+#include <wtf/text/TextStream.h>
```
Does `operator<<` need to be inlined? If not, we can avoid including `TextStream.h` here, and use a forward declaration instead; this is good to resist.

nt1m replied:
It really doesn't, will move to .cpp

darinadler commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -1271,9 +1271,22 @@ bool SelectorChecker::checkOne(CheckingContext& checkingContext, LocalContext& c
 
-            // Wildcard always matches.
+            // Universal selector always matches.
             auto& argument = selector.argumentList()->first();
```
Sure seems like this would read nicer if we did this, since we use it 4 times:
```
auto& list = *selector.argumentList();
```

darinadler commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -1271,9 +1271,22 @@ bool SelectorChecker::checkOne(CheckingContext& checkingContext, LocalContext& c
+                return true;
+
+            auto result = std::ranges::all_of(
+                selector.argumentList()->begin() + 1, selector.argumentList()->end(),
```
Would read so much more nicely with the local variable and less wrapping:
```
auto result = std::ranges::all_of(list.begin() + 1, list.end(),
```

nt1m replied:
Good suggestion!

darinadler reviewed (approved):
Does this extra data have measurable performance cost? I suppose we’ll find out from our benchmarks.

darinadler commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -377,6 +377,11 @@ static void appendPseudoClassFunctionTail(StringBuilder& builder, const CSSSelec
 
 }
 
+inline WTF::TextStream& operator<<(WTF::TextStream& ts, PossiblyQuotedIdentifier quoted)
```
Need to remove the `inline` here.

nt1m committed:
[view-transitions] Implement matching for view transition classes
https://bugs.webkit.org/show_bug.cgi?id=278223
rdar://134020027

Reviewed by Darin Adler.

Implements matching as described in:
https://drafts.csswg.org/css-view-transitions-2/#shared-style-with-vt-classes

* LayoutTests/TestExpectations:

Unskip all tests.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:

Enable by default, since all WPT pass.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::operator<<):
* Source/WebCore/css/CSSSelector.h:

Add PossiblyQuotedIdentifier TextStream for easy debugging.

* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorChecker.h:

All classes in need to match in order for the styles to apply.

* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::effectiveViewTransitionClassList):
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::captureNewState):

Captures the classList as described in https://drafts.csswg.org/css-view-transitions-2/#vt-class-algorithms

* Source/WebCore/dom/ViewTransition.h: Add classList to CapturedElement.

* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::pseudoStyleCacheIsInvalid):

Fix a bug in the cache where were passing a null name argument string.

* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::classListForNamedViewTransitionPseudoElement):
(WebCore::Style::ElementRuleCollector::ruleMatches):

Pass the classList to the SelectorChecker when checking the selector.

Canonical link: https://commits.webkit.org/282383@main
